### ISSUE TITLE
the error of grad_scale

### ISSUE THREAD

**Issue by wuyeyexvnainai** [zhanyisun/lpb#2]
<img width="896" height="636" alt="Image" src="https://github.com/user-attachments/assets/650336af-9eb3-4e5c-b87a-adbe572872cc" />

`        for t in scheduler.timesteps:
            # 1. apply conditioning
            trajectory[condition_mask] = condition_data[condition_mask]
            trajectory = trajectory.detach().requires_grad_()

            # 2. predict model output 
            model_output = model(trajectory, t, 
                local_cond=local_cond, global_cond=global_cond)

            # classifier guidance
            if classifier_guidance and t < self.guidance_start_timestep and current_cost > self.threshold:
                trajectory0 = scheduler.step(model_output, t, trajectory).pred_original_sample # x_0
                loss = self.planner.compute_loss(trajectory0, current_obs)
                cond_grad = -torch.autograd.grad(loss, trajectory)[0]
                guidance_scale = self.guidance_scale
                grad_scale = guidance_scale * (1 - scheduler.alphas_cumprod[t]).sqrt() 
                trajectory = trajectory.detach() + grad_scale * cond_grad

            # 3. compute previous image: x_t -> x_t-1
            trajectory = scheduler.step(
                model_output, t, trajectory, 
                generator=generator,
                **kwargs
                ).prev_sample`

Should the sqrt() here be removed
Because later on, modifications were made to the trajectory rather than the model_output

**Comment by zhanyisun** [comment]
Hi! Thanks for the question!

According to the original paper on classifier guidance https://arxiv.org/pdf/2105.05233, in algorithm 2, there's an sqrt operation for the coefficient in front of the gradient. Let me know if you have additional questions!

**Comment by wuyeyexvnainai** [comment]
Hello, author
The formula in the text is for noise "model_ouput", but here it is for trajectory. If it is for noise, it does need to be marked with a root, but for trajectory, does it not need to be marked with a root

<img width="677" height="64" alt="Image" src="https://github.com/user-attachments/assets/0a5128ff-1da1-4663-a20e-9796e1fc7ddf" />

<img width="770" height="418" alt="Image" src="https://github.com/user-attachments/assets/f7c75982-e186-495a-8f06-012163c6a8c7" />

**Comment by zhanyisun** [comment]
Hi!

Thanks for the follow up! Start from the standard ε-space guidance from the paper

<img width="373" height="54" alt="Image" src="https://github.com/user-attachments/assets/ef1f91f8-a0fa-4c31-9e9a-11b65b863e3f" />

And x_t and \varepsilon are linked by

<img width="404" height="61" alt="Image" src="https://github.com/user-attachments/assets/6ded21ba-ce55-4705-85cc-ecedfcf1b424" />

A small change in noise \Delta\varepsilon induces a change in the sample of size

<img width="208" height="44" alt="Image" src="https://github.com/user-attachments/assets/ea29921d-d16a-487f-bee8-4421e810acbe" />

Therefore, to get the same effect while updating the trajectory x_t directly, you should scale the x-update by the same \sigma_t

<img width="398" height="46" alt="Image" src="https://github.com/user-attachments/assets/4dad4a57-0bc5-4452-a457-f38f73c9f12d" />

**Comment by wuyeyexvnainai** [comment]
hello author.
Thank you very much for your answer
But i still have some questions
<img width="408" height="693" alt="Image" src="https://github.com/user-attachments/assets/f61c367e-6b39-4d1d-bca3-7bb9e85df0d3" />

**Comment by zhanyisun** [comment]
Thanks for the follow up! I think you are right - good catch! 
The line should be `grad_scale = guidance_scale * (1 - scheduler.alphas_cumprod[t])` to match exactly with the math. Empirically I have found having the .sqrt() in the coefficient also works up to a point if guidance_scale is well-tuned. Since 0≤1-alphas_cumprod[t]≤1, the square-root schedule is larger early (when noise is high) and smaller late (when near-denoised). So the code is probably pushing harder at earlier, noisier steps and under-guide late steps compared to the original paper. In practice, guidance strength is a tunable hyperparam so I guess one explanation is that the guidance_scale likely absorbed most of this coefficient mismatch already.
